<font size="6"> <div class="alert alert-block alert-info"><center>**Segundo trabalho individual de Otimização Heurística**</center></div></font>

<center><h4> Trabalho realizado por Sebastião Rosalino, nº98437 turma CDB1, Licenciatura de Ciência de Dados</h4></center>

<center><h4> Docente: Anabela Costa </h4></center>

<center><img src="Caixeiro_profile.png" width=250px height=150px></center>

<center><img src="logo_iscte.png" width=200px height=200px></center>

#### Importação das bibliotecas necessárias

In [7]:
import pandas as pd
import random
import numpy as np

<strong> Nota: Na resolução das várias alíneas do trabalho, assume-se que, no âmbito do Problema do Caixeiro Viajante, todos os percursos foram realizados seguindo o paradigma de colocar como última cidade visitada a cidade inicial do percurso.
</strong>

## **a) Determine uma solução admissível para o problema recorrendo à Heurística do Vizinho mais Próximo**

Neste problema, determinar uma solução admissível consiste em encontrar um percurso válido com uma distância percorrida associada.

* **Etapa 1**:

Construção do ficheiro “Distancias.xlsx” com a instância do Problema do Caixeiro Viajante em questão, com o objetivo de utilização nos algoritmos usados na resolução do problema.

* **Etapa 2**:

Leitura e visualização em código python do ficheiro criado na etapa 1, conforme descrito abaixo:

In [8]:
mapa_caixeiro = pd.read_excel('Distancias.xlsx')

mapa_caixeiro.index += 1

mapa_caixeiro

mapa_caixeiro.fillna('-')

1     2     3     4     5     6     7     8
1     -  14.0  15.0     -     -     -     -  17.0
2  14.0     -  13.0  14.0  20.0     -     -  21.0
3  15.0  13.0     -  11.0  21.0  17.0   9.0   9.0
4     -  14.0  11.0     -  11.0  10.0   8.0  20.0
5     -  20.0  21.0  11.0     -  15.0  18.0     -
6     -     -  17.0  10.0  15.0     -   9.0     -
7     -     -   9.0   8.0  18.0   9.0     -  13.0
8  17.0  21.0   9.0  20.0     -     -  13.0     -

* **Etapa 3**:

Devido à falta de ligações entre algumas cidades, consoante a cidade inicial pode ou não ser possível completar o percurso. 
Optou-se, assim, por iniciar o percurso na cidade 3 (K1 = 3), com o objetivo de percorrer todas as cidades.

* **Etapa 4**:

Procedeu-se, de seguida, à aplicação da Heurística Construtiva do “Vizinho mais Próximo”, cuja abordagem passa por visitar sempre a cidade (vértice) mais próxima da mais recente, desde que não tenha sido visitada previamente, de forma a procurar uma rota de percurso admissível.
Foi usado para o efeito o seguinte código:

In [9]:
# Leitura do ficheiro excel que tem a tabela das distâncias entre as cidades consideradas na instância do Problema 
# do Caixeiro Viajante

distance_matrix = mapa_caixeiro

nodes = len(distance_matrix) # A função len() indica o tamanho do objeto

# Indicação da cidade onde começa o circuito (X0)
# Exemplo: se X0=3 então o circuito é iniciado na cidade 3

x0 = 3
if not x0:
    x0 = random.randint(1, nodes)

path = [x0] # lista que tem a sequência de cidades que o caixeiro Viajante irá visitar

distance = 0

finish = True
while finish:
    closest_neighbour = distance_matrix.loc[path[-1], ~distance_matrix.index.isin(path)].idxmin()
    distance += distance_matrix.loc[path[-1], closest_neighbour]
    path.append(closest_neighbour)
    if len(path) == nodes:
        distance += distance_matrix.loc[path[-1], x0]
        finish = False
path.append(x0) # Adição do caminho de volta à cidade inicial ao circuito        
    
print('Distância total do Circuito Gerado:', distance, 'unidades') # Distância do circuito do Caixeiro Viajante
print()
print('Circuito gerado para o Caixeiro Viajante:')
print(path) # Circuito gerado para o Caixeiro Viajante

Distância total do Circuito Gerado: 102.0 unidades

Circuito gerado para o Caixeiro Viajante:
[3, 7, 4, 6, 5, 2, 1, 8, 3]


* **Etapa 5**:

Aplicado o código, obteve-se como solução admissível o seguinte percurso:

Para K1 = 3, temos:

A = {(3,7), (7,4), (4,6), (6,5), (5,2), (2,1), (1,8), (8,3)}

Circuito: [3, 7, 4, 6, 5, 2, 1, 8, 3]

Sendo A a solução obtida.

A distância total deste circuito é igual a 102 = 9 + 8 + 10 + 15 + 20 + 14 + 17 + 9

Comentário à solução obtida:

No circuito obtido, as distâncias aumentam com algum significado após os três primeiros percursos (a partir da cidade 6), embora se volte a reduzir na última ligação (8,3). No entanto, não é possível, apenas com a informação disponível, formular uma avaliação sobre o nível de eficiência desta solução. Para tal seria necessário avaliar soluções alternativas. 

Para tentar obter uma solução diferente (eventualmente com menor distância), poderia 
repetir-se a aplicação da heurística, mas iniciando o circuito a partir de uma outra cidade.

<h2><b> b) Considere a seguinte Estrutura de Vizinhança : </b></h2>

<ul><li><h3> A solução S1 é vizinha da solução S se é obtida a partir de S através da troca de posição, no circuito, de dois vértices/ cidades. </h3></li></ul>

Uma Estrutura de Vizinhança especifica quais as soluções que são vizinhas (alcançáveis numa única iteração) de qualquer solução atual.

Uma solução vizinha da solução atual é aquela que é alcançada através de uma operação 2–optimal (i.e., pela troca de posição, no circuito, de dois vértices/cidades, levando à obtenção de um novo circuito). 

## **b1) Determine uma solução vizinha da solução apresentada em a).**

Pode considerar-se esta questão no domínio das Heurísticas Melhorativas, como sendo as que partem duma solução e, a cada passo, introduzem pequenas alterações à solução de forma a tentar obter uma solução melhor. 

No caso em análise, dada uma solução S para o TSP, uma solução S1 pertence à vizinhança de S se for obtida a partir de S, trocando as posições de 2 cidades/vértices no percurso (o equivalente a remover 2 arestas não consecutivas e inserir 2 novas arestas que garantam a obtenção de um novo circuito).


**Regra**:

Sejam (i, j) e (k, l) , com j diferente de k, as arestas removidas. 
Então, as arestas a serem inseridas são (i, k) e (j, l).

A solução S (solução atual) é a seguinte:

{(3,7), (7,4), (4,6), (6,5), (5,2), (2,1), (1,8), (8,3)}

Circuito: [3, 7, 4, 6, 5, 2, 1, 8, 3]

Considerando o conceito de vizinhança acima referido e de acordo com o solicitado, proceder-se-á à determinação de uma solução vizinha manualmente.

Para o efeito, decidiu-se retirar as arestas (7,4) e (6,5) e inserir as arestas (7,6) e (4,5), o que significa a troca de posição no circuito da cidade 4 com a 6.

Sendo um TSP simétrico, a ligação entre as cidades 4 e 6 será invertida no novo circuito.

A nova solução S1 (vizinha de S) será a seguinte:

{(3,7), (7,6), (6,4), (4,5), (5,2), (2,1), (1,8), (8,3)}

Circuito: [3, 7, 6, 4, 5, 2, 1, 8, 3]

<table>
    <tr>
        <th> Ligações Removidas </th>
        <th> Novas Ligações </th>
        <th> Cidades Trocadas </th>
        <th> Circuito </th>
        <th> Distância da Solução </th>
    </tr>
    <tr>
        <td> (7,4), (6,5) </td>
        <td> (7,6), (4,5) </td>
        <td> 4 e 6 </td>
        <td> [3, 7, 6, 4, 5, 2, 1, 8, 3]
        <td> 99 unidades </td>
    </tr>
</table>

Explicações adicionais:

* Esta solução vizinha só é admissível porque existe ligação entre a cidade 7 e 6, e da cidade 4 para a 5.

* À distância total remove-se a ligação entre as cidades 7 e 4 (distância -8) e a ligação entre as cidades 6 e 5 (distância -15).

* Adiciona-se à distância total a ligação entre as cidades 7 e 6 (distância +9) e a ligação entre as cidades 4 e 5 (distância +11).

* A ligação (4,6) é invertida para (6,4), o que não altera a dimensão do percurso porque estamos em presença de um problema de natureza simétrica;

* Assim, a solução S1 encontrada é definida pelo percurso [3, 7, 6, 4, 5, 2, 1, 8, 3] com distância percorrida de 99 unidades, o que representa uma melhoria face à solução inicial S.

## **b2) Defina uma lista Tabu.**

Usando a metodologia acima descrita, existe a possibilidade de se selecionar soluções vizinhas já analisadas. Para precaver esta situação, utiliza-se um mecanismo, que é designado por Lista Tabu, que evita o retorno a soluções anteriores. 

Por definição, em cada iteração do algoritmo adiciona-se à Lista Tabu o par de cidades trocadas que dá origem à solução vizinha obtida (movimentos tabu).

Para o caso em análise definir-se-á a Lista Tabu como uma lista de dimensão 1 contendo o par de cidades trocadas na obtenção da solução vizinha.
Se esta definição de Lista Tabu tivesse sido aplicada da passagem da solução inicial S para a solução vizinha S1 obtida em b1), ter-lhe-ia sido adicionado o par de cidades 4 e 6.

Lista Tabu: cidades [4, 6]

<h2><b> b3) Realize uma iteração do Algoritmo de Pesquisa Tabu, assumindo que: </b></h2>

<ul>
    <li><h4> A Solução Atual corresponde à solução vizinha gerada na alínea b1); </h4></li>
    <li><h4> A lista Tabu está vazia </h4></li>
</ul>

O Algoritmo de Pesquisa Tabu pode ser definido como um método de procura de uma nova solução (não necessariamente melhor do que a anterior) na vizinhança da Solução Atual – de modo iterativo – com vista a obter uma solução global do problema (ótimo global).

Este método inclui um Procedimento de Pesquisa Local aplicado dentro de um determinado espaço alargado de procura, que permite determinar a melhor solução vizinha. Contudo, na Pesquisa Tabu a melhor solução vizinha pode ser uma solução pior do que a que lhe deu origem.

Esta possibilidade de selecionar soluções vizinhas piores, cria um problema: a possibilidade de haver retorno a uma solução anterior. Para evitar esta situação, a Pesquisa Tabu utiliza um mecanismo, que é designado por Lista Tabu, que evita o retorno a soluções anteriores.

<center><h3><b> Dados iniciais - Inicialização </b></h3></center>
<br>
<table>
    <tr>
        <th> Solução Atual </th>
        <th> Circuito </th>
        <th> Distância percorrida </th>
        <th> Lista Tabu </th>
    </tr>
    <tr>
        <td> {(3,7), (7,6), (6,4), (4,5), (5,2), (2,1), (1,8), (8,3)} </td>
        <td> [3, 7, 6, 4, 5, 2, 1, 8, 3]
        <td> 99 unidades </td>
        <td> { } </td> 
     </tr>
</table>

* **Iteração 1**:

Para aplicação da primeira iteração aplicou-se o seguinte código:

In [10]:
# Função chunks gera sequências sucessivas de tamanho n a partir de uma lista 

def chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    mini_paths = pd.DataFrame(columns=['begin', 'end'])
    for i in range(0, len(lst)-1, n):
        mini_paths.loc[i, :] = lst[i:i + 2]
    return mini_paths

# Leitura da matriz de distâncias da instância do TSP 

distance_matrix = mapa_caixeiro

nodes = len(distance_matrix)

path = [3, 7, 6, 4, 5, 2, 1, 8, 3]  # lista com o caminho correspondente à solução Atual

distance = 99 # Distância associada à solução Atual
        
# O circuito inicial para o TSP encontra-se no path e a distância do circuito está na distance        

# Aplicação do Procedimento 2-optimal (a partir da solução inicial do TSP)

# Geração das arestas não consecutivas do circuito inicial do TSP que irão conduzir a novos circuitos para o problema

mini_paths = chunks(path, 1)

# Geração das soluções/ circuitos vizinhas da solução inicial do TSP 

all_paths = pd.DataFrame(columns=['initial_path', 'new_path', 'path_lst', 'new_distance'])

contador = 0
best_distance = distance
best_path = path
for i, ri in mini_paths.iterrows():
    filter_adjacent = np.logical_and(~mini_paths['end'].isin([ri['begin'], ri['end']]),
                                     ~mini_paths['begin'].isin([ri['begin'], ri['end']]))
    filter_simetrical = mini_paths.index >= i
    filter_lines = np.logical_and(filter_adjacent, filter_simetrical)

    it_paths = mini_paths.loc[filter_lines, :]

    for j, rj in it_paths.iterrows():
        new_path_lst = path.copy()

        distance_it = distance - distance_matrix.at[int(ri['begin']), int(ri['end'])] - distance_matrix.at[int(rj['begin']), int(rj['end'])]
        distance_it += distance_matrix.at[int(ri['begin']), int(rj['begin'])] + distance_matrix.at[int(ri['end']), int(rj['end'])]

        initial_path = f"({int(ri['begin'])}, {int(ri['end'])}) ({int(rj['begin'])}, {int(rj['end'])})"
        new_path = f"({int(ri['begin'])}, {int(rj['begin'])}) ({int(ri['end'])}, {int(rj['end'])})"

        pos_initial = path.index(ri['end'])
        pos_new = path.index(rj['begin'])
        new_path_lst[pos_initial], new_path_lst[pos_new] = new_path_lst[pos_new], new_path_lst[pos_initial]

        all_paths.loc[contador, :] = {'initial_path': initial_path, 'new_path': new_path, 'path_lst': new_path_lst,
                                     'new_distance': distance_it}
        
        if distance_it < best_distance:
            best_distance = distance_it
            best_path = new_path_lst

        contador += 1

In [11]:
all_paths.index += 1
all_paths.columns = ['Ligações removidas', 'Ligações adicionadas', 'Novo Circuito', 'Distância Percorrida']
all_paths['Distância Percorrida'] = all_paths['Distância Percorrida'].fillna('Não admissível')

def cidades_trocadas(ligacoes):
    return f'{ligacoes[4]} e {ligacoes[8]}'

all_paths['Cidades trocadas'] = all_paths['Ligações removidas'].apply(cidades_trocadas)

all_paths = all_paths[['Ligações removidas', 'Ligações adicionadas', 'Cidades trocadas', 'Novo Circuito', 'Distância Percorrida']]

### Lista de todas as soluções vizinhas da solução atual

In [12]:
all_paths

Ligações removidas Ligações adicionadas Cidades trocadas  \
1       (3, 7) (6, 4)        (3, 6) (7, 4)            7 e 6   
2       (3, 7) (4, 5)        (3, 4) (7, 5)            7 e 4   
3       (3, 7) (5, 2)        (3, 5) (7, 2)            7 e 5   
4       (3, 7) (2, 1)        (3, 2) (7, 1)            7 e 2   
5       (3, 7) (1, 8)        (3, 1) (7, 8)            7 e 1   
6       (7, 6) (4, 5)        (7, 4) (6, 5)            6 e 4   
7       (7, 6) (5, 2)        (7, 5) (6, 2)            6 e 5   
8       (7, 6) (2, 1)        (7, 2) (6, 1)            6 e 2   
9       (7, 6) (1, 8)        (7, 1) (6, 8)            6 e 1   
10      (7, 6) (8, 3)        (7, 8) (6, 3)            6 e 8   
11      (6, 4) (5, 2)        (6, 5) (4, 2)            4 e 5   
12      (6, 4) (2, 1)        (6, 2) (4, 1)            4 e 2   
13      (6, 4) (1, 8)        (6, 1) (4, 8)            4 e 1   
14      (6, 4) (8, 3)        (6, 8) (4, 3)            4 e 8   
15      (4, 5) (2, 1)        (4, 2) (5, 1)            5 e 2   
16      (4, 5) (1, 8)        (4, 1) (5, 8)            5 e 1   
17      (4, 5) (8, 3)        (4, 8) (5, 3)            5 e 8   
18      (5, 2) (1, 8)        (5, 1) (2, 8)            2 e 1   
19      (5, 2) (8, 3)        (5, 8) (2, 3)            2 e 8   
20      (2, 1) (8, 3)        (2, 8) (1, 3)            1 e 8   

                  Novo Circuito Distância Percorrida  
1   [3, 6, 7, 4, 5, 2, 1, 8, 3]                105.0  
2   [3, 4, 6, 7, 5, 2, 1, 8, 3]                108.0  
3   [3, 5, 6, 4, 7, 2, 1, 8, 3]       Não admissível  
4   [3, 2, 6, 4, 5, 7, 1, 8, 3]       Não admissível  
5   [3, 1, 6, 4, 5, 2, 7, 8, 3]                101.0  
6   [3, 7, 4, 6, 5, 2, 1, 8, 3]                102.0  
7   [3, 7, 5, 4, 6, 2, 1, 8, 3]       Não admissível  
8   [3, 7, 2, 4, 5, 6, 1, 8, 3]       Não admissível  
9   [3, 7, 1, 4, 5, 2, 6, 8, 3]       Não admissível  
10  [3, 7, 8, 4, 5, 2, 1, 6, 3]                111.0  
11  [3, 7, 6, 5, 4, 2, 1, 8, 3]                 98.0  
12  [3, 7, 6, 2, 5, 4, 1, 8, 3]       Não admissível  
13  [3, 7, 6, 1, 5, 2, 4, 8, 3]       Não admissível  
14  [3, 7, 6, 8, 5, 2, 1, 4, 3]       Não admissível  
15  [3, 7, 6, 4, 2, 5, 1, 8, 3]       Não admissível  
16  [3, 7, 6, 4, 1, 2, 5, 8, 3]       Não admissível  
17  [3, 7, 6, 4, 8, 2, 1, 5, 3]                120.0  
18  [3, 7, 6, 4, 5, 1, 2, 8, 3]       Não admissível  
19  [3, 7, 6, 4, 5, 8, 1, 2, 3]       Não admissível  
20  [3, 7, 6, 4, 5, 2, 8, 1, 3]                112.0

Análises às soluções obtidas:

* Após uma iteração do Algoritmo de Pesquisa Tabu, a **Melhor Solução Vizinha** encontrada corresponde à solução 11: 

    {(3,7), (7,6), (6,5), (5,4), (4,2), (2,1), (1,8), (8,3)} com o seguinte circuito: [3, 7, 6, 5, 4, 2, 1, 8, 3], com uma distância associada de 98 unidades. 

* A **Melhor Solução Conhecida** neste caso, como a distância de circuito diminuiu (passou de 99 para 98 unidades de distância), é atualizada para a solução correspondente à Melhor Solução Vizinha, ou seja, a solução {(3,7), (7,6), (6,5), (5,4), (4,2), (2,1), (1,8), (8,3)}, de circuito associado:

    [3, 7, 6, 5, 4, 2, 1, 8, 3].

* A **Lista Tabu** sofrerá alterações, deixando de estar vazia. Ser-lhe-á acrescentado o par de cidades trocadas na obtenção da Melhor Solução Vizinha, ou seja o par de cidades 4 e 5.

* Um **Movimento Tabu** é gerado quando um movimento de reposicionamento no circuito é efetuado para um determinado par de cidades. Ou seja, quando duas cidades são escolhidas para serem trocadas de posição no circuito, o movimento tabu a gerar visa proibir que essas cidades possam voltar a ser trocadas entre si. No caso em análise, o **Movimento Tabu** pode ser definido como o movimento que conduziu à troca de posição das cidades 4 e 5 no percurso e que passou a integrar da lista tabu. 

    Movimento Tabu: corresponde à troca das cidades 4 e 5. 

* A **Solução Atual**, numa seguinte iteração, passaria a ser a Melhor Solução Vizinha da iteração efetuada, portanto a solução:

    {(3,7), (7,6), (6,5), (5,4), (4,2), (2,1), (1,8), (8,3)} de circuito: [3, 7, 6, 5, 4, 2, 1, 8, 3].

<center><h3><b> Dados após uma iteração (iteração 1) </b></h3></center>
<br>
<table>
    <tr>
        <th> Solução Atual </th>
        <th> Circuito </th>
        <th> Distância percorrida </th>
        <th> Lista Tabu </th>
    </tr>
    <tr>
        <td> {(3,7), (7,6), (6,5), (5,4), (4,2), (2,1), (1,8), (8,3)} </td>
        <td> [3, 7, 6, 5, 4, 2, 1, 8, 3]
        <td> 98 unidades </td>
        <td> { cidades 4 e 5 } </td> 
     </tr>
</table>

Se o algoritmo continuasse a solução atual seria atualizada em todas as iterações, no entanto a melhor solução conhecida só seria atualizada caso seja encontrada uma solução com uma distância percorrida associada menor.